--- drawing/nx_pylab.py~	2008-01-24 23:21:52.000000000 +0000
+++ drawing/nx_pylab.py	2008-01-24 23:25:02.000000000 +0000
@@ -176,4 +176,5 @@
                         vmax=None, 
                         ax=None,
+                        linewidths=None,
                         **kwds):
     """Draw nodes of graph G
@@ -215,5 +216,6 @@
                                vmin=vmin,
                                vmax=vmax,
-                               alpha=alpha)
+                               alpha=alpha,
+                               linewidths=linewidths)
                                
     node_collection.set_zorder(2)            
